File: src/lib/research/keywords.ts
/**
 * 中日市場のキーワード差分析（Phase 2、docs/development_plan.md セクション 41）。
 *
 * backend/app/keywords.py と同一の抽出・集計を移植し、API 未接続のモックでも同じ
 * キーワード差を返す（整合性の原則: セクション 93）。文字 n-gram はコードポイント単位で
 * 一致するため、日本語（BMP）で Python と結果が揃う。
 */

import { pyRound } from "./history";
import type { KeywordBias, KeywordGap, MarketSnapshot, ProductCatalogEntry } from "./types";

const MIN_LEN = 3;
const MAX_LEN = 5;
const MIN_DOC_FREQ = 2;
const BIAS_THRESHOLD = 8;
const MAX_KEYWORDS = 12;

function clean(name: string): string {
  return name.replace(/ /g, "").replace(/　/g, "");
}

function substrings(name: string): Set<string> {
  const cleaned = clean(name);
  const result = new Set<string>();
  for (let size = MIN_LEN; size <= MAX_LEN; size++) {
    for (let start = 0; start <= cleaned.length - size; start++) {
      result.add(cleaned.slice(start, start + size));
    }
  }
  return result;
}

function documentFrequencies(entries: ProductCatalogEntry[]): Map<string, number> {
  const freq = new Map<string, number>();
  for (const entry of entries) {
    for (const sub of substrings(entry.name)) {
      freq.set(sub, (freq.get(sub) ?? 0) + 1);
    }
  }
  return freq;
}

/** 文字列の辞書順比較（-1/0/1）。 */
function compareStr(a: string, b: string): number {
  if (a < b) return -1;
  if (a > b) return 1;
  return 0;
}

function extractKeywords(entries: ProductCatalogEntry[]): string[] {
  const freq = documentFrequencies(entries);
  const candidates = [...freq.entries()].filter(([, count]) => count >= MIN_DOC_FREQ).map(([sub]) => sub);
  // 長い順・辞書順で走査し、同頻度でより長い採用済み語に含まれる短い語は捨てる。
  candidates.sort((a, b) => (a.length !== b.length ? b.length - a.length : compareStr(a, b)));
  const kept: string[] = [];
  for (const candidate of candidates) {
    const redundant = kept.some(
      (k) => k.includes(candidate) && freq.get(k) === freq.get(candidate) && k.length > candidate.length,
    );
    if (!redundant) kept.push(candidate);
  }
  return kept;
}

function sideStrength(snap: MarketSnapshot): number {
  const reviewScore = Math.min(100, snap.reviewCount / 20);
  const competitionPenalty = Math.min(30, snap.competitors * 0.15);
  const raw = snap.demandIndex * 0.7 + reviewScore * 0.3 - competitionPenalty;
  return Math.max(0, Math.min(100, raw));
}

function mean(values: number[]): number {
  return values.length ? values.reduce((a, b) => a + b, 0) / values.length : 0;
}

function biasOf(gap: number): KeywordBias {
  if (gap >= BIAS_THRESHOLD) return "jp";
  if (gap <= -BIAS_THRESHOLD) return "cn";
  return "balanced";
}

/** キーワードごとの日中市場強度差を、差の大きい順に返す。 */
export function keywordGaps(entries: ProductCatalogEntry[]): KeywordGap[] {
  const keywords = extractKeywords(entries);
  const gaps: KeywordGap[] = [];
  for (const keyword of keywords) {
    const matched = entries.filter((e) => clean(e.name).includes(keyword));
    if (matched.length === 0) continue;
    const jpStrength = pyRound(mean(matched.map((e) => sideStrength(e.japan))));
    const cnStrength = pyRound(mean(matched.map((e) => sideStrength(e.china))));
    const gap = jpStrength - cnStrength;
    gaps.push({ keyword, productCount: matched.length, jpStrength, cnStrength, gap, bias: biasOf(gap) });
  }

  gaps.sort((a, b) => {
    if (Math.abs(a.gap) !== Math.abs(b.gap)) return Math.abs(b.gap) - Math.abs(a.gap);
    if (a.productCount !== b.productCount) return b.productCount - a.productCount;
    return compareStr(a.keyword, b.keyword);
  });
  return gaps.slice(0, MAX_KEYWORDS);
}

File: backend/app/keywords.py
"""中日市場のキーワード差分析（Phase 2, docs/development_plan.md セクション 41）。

商品名から再出現する文字 n-gram をキーワードとして抽出し、そのキーワードを含む商品群の
日本市場・中国市場の「市場強度」（需要・レビュー・競合から算出）を比較する。差が大きい
キーワードは、片側の市場で強く他方で手薄＝商機を示す。決定論的に確定し（原則: セクション 93）、
生成式はフロント（keywords.ts）と一致させる。
"""

from __future__ import annotations

from statistics import mean

from .catalog import CatalogEntry
from .schemas import KeywordGap, MarketSnapshot

# 抽出する n-gram の長さ範囲と、再出現とみなす最小文書頻度。
# 2 文字は語境界をまたぐ断片ノイズ（カー・ンタ等）が多いため 3 文字以上とする。
_MIN_LEN = 3
_MAX_LEN = 5
_MIN_DOC_FREQ = 2
# バイアス判定のしきい値（強度差）。
_BIAS_THRESHOLD = 8
# 返すキーワードの上限。
_MAX_KEYWORDS = 12


def _substrings(name: str) -> set[str]:
    """名前から長さ _MIN_LEN.._MAX_LEN の部分文字列集合を返す。"""
    cleaned = name.replace(" ", "").replace("　", "")
    result: set[str] = set()
    length = len(cleaned)
    for size in range(_MIN_LEN, _MAX_LEN + 1):
        for start in range(length - size + 1):
            result.add(cleaned[start : start + size])
    return result


def _document_frequencies(entries: list[CatalogEntry]) -> dict[str, int]:
    freq: dict[str, int] = {}
    for entry in entries:
        for sub in _substrings(entry.name):
            freq[sub] = freq.get(sub, 0) + 1
    return freq


def _extract_keywords(entries: list[CatalogEntry]) -> list[str]:
    """再出現する n-gram のうち、より長い同頻度語に包含されない代表語を返す。"""
    freq = _document_frequencies(entries)
    candidates = [sub for sub, count in freq.items() if count >= _MIN_DOC_FREQ]
    # 長い順・辞書順で走査し、同頻度でより長い採用済み語に含まれる短い語は捨てる。
    candidates.sort(key=lambda s: (-len(s), s))
    kept: list[str] = []
    for candidate in candidates:
        redundant = any(candidate in k and freq[k] == freq[candidate] and len(k) > len(candidate) for k in kept)
        if not redundant:
            kept.append(candidate)
    return kept


def _side_strength(snap: MarketSnapshot) -> float:
    """1 市場の強度（0-100）。需要中心、レビューで加点、競合で減点。"""
    review_score = min(100.0, snap.review_count / 20)
    competition_penalty = min(30.0, snap.competitors * 0.15)
    raw = snap.demand_index * 0.7 + review_score * 0.3 - competition_penalty
    return max(0.0, min(100.0, raw))


def _bias(gap: int) -> str:
    if gap >= _BIAS_THRESHOLD:
        return "jp"
    if gap <= -_BIAS_THRESHOLD:
        return "cn"
    return "balanced"


def keyword_gaps(entries: list[CatalogEntry]) -> list[KeywordGap]:
    """キーワードごとの日中市場強度差を、差の大きい順に返す。"""
    keywords = _extract_keywords(entries)
    gaps: list[KeywordGap] = []
    for keyword in keywords:
        matched = [e for e in entries if keyword in e.name.replace(" ", "").replace("　", "")]
        if not matched:
            continue
        jp_strength = round(mean(_side_strength(e.japan) for e in matched))
        cn_strength = round(mean(_side_strength(e.china) for e in matched))
        gap = jp_strength - cn_strength
        gaps.append(
            KeywordGap(
                keyword=keyword,
                product_count=len(matched),
                jp_strength=jp_strength,
                cn_strength=cn_strength,
                gap=gap,
                bias=_bias(gap),
            )
        )

    gaps.sort(key=lambda g: (-abs(g.gap), -g.product_count, g.keyword))
    return gaps[:_MAX_KEYWORDS]
